File: src-tauri/src/crawler/alibaba.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use anyhow::Error;
use serde_json::from_str;

use crate::constant::TranslateQuery;

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
struct CsrfToken {
  token: String,
  parameter_name: String,
  header_name: String,
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Response {
    code: String,
    data: Data,
    http_status_code: i64,
    message: String,
    request_id: String,
    success: bool,
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Data {
    detect_language: String,
    translate_text: String,
}

pub async fn translate(query: TranslateQuery) -> Result<String, Error> {
  // println!("Alibaba");
  // 获取csrftoken "{\"token\":\"136fd918-58a1-4d14-90b3-254cc15f9af7\",\"parameterName\":\"_csrf\",\"headerName\":\"X-XSRF-TOKEN_PROPERTY_ITEM\"}"
  let user_agent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36".to_string();
  let client = reqwest::Client::builder().build()?;
  let res = client.get("https://translate.alibaba.com/api/translate/csrftoken").
    header("User-Agent", &user_agent).
    send().await?.text().await?;
  let csrf_token: CsrfToken = from_str(&res)?;
  // print!("==== csrfToken: {:?}", csrf_token);

  // 构建请求的数据
  let token = csrf_token.token;
  let mut params = HashMap::new();
  params.insert("srcLang", query.source);
  params.insert("tgtLang", query.target);
  params.insert("domain", "general".to_string());
  params.insert("query", query.text);
  params.insert(&csrf_token.parameter_name, token.to_string());
  // println!("====== form_dat: {:?}", params);

  // 发送请求
  let res2 =
    client.post("https://translate.alibaba.com/api/translate/text").
    header("User-Agent", &user_agent).
    header(csrf_token.header_name, token.to_string()).
    header("Content-Type", "multipart/form-data").
    form(&params)
    .send().await?.text().await?;
  println!("===== result {:?}", res2);

  let data: Response = from_str(&res2)?;

  Ok(data.data.translate_text)
}

File: src/services/translate/alibaba.ts
import { ITranslateParams, TLanguage } from '@/common/constants';
import { ResponseType, getClient, Body } from '@tauri-apps/api/http';
import type { Client } from '@tauri-apps/api/http';

interface CsrfToken {
  token: string;
  parameterName: string;
  headerName: string;
}

interface ITranslateData {
  code: string;
  data: {
    detectLanguage: TLanguage;
    translateText: string;
  };
  httpStatusCode: number;
  message: string;
  requestId: string;
  success: boolean;
}

let _client: Client;
const USER_AGENT =
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36';

async function initClient(): Promise<Client> {
  if (_client) return _client;
  _client = await getClient();
  return _client;
}

export async function translate(params: ITranslateParams): Promise<string> {
  try {
    const client = await initClient();
    /**
     * 获取 csrftoken
     */
    const res1 = await client.get<CsrfToken>('https://translate.alibaba.com/api/translate/csrftoken', {
      headers: {
        'User-Agent': USER_AGENT,
      },
      responseType: ResponseType.JSON,
    });
    // console.log('csrfToken：', res1.data);
    const { token, parameterName, headerName } = res1.data;
    /**
     * 发起翻译请求
     */
    const form = new FormData();
    form.append('srcLang', params.source);
    form.append('tgtLang', params.target);
    form.append('domain', 'general');
    form.append('query', params.text);
    form.append(parameterName, token);
    const res = await client.post<ITranslateData>('https://translate.alibaba.com/api/translate/text', Body.form(form), {
      headers: {
        'User-Agent': USER_AGENT,
        [headerName]: token,
        'Content-Type': 'multipart/form-data',
      },
      responseType: ResponseType.JSON,
    });
    // console.log('翻译后的内容：', res);
    if (res.data?.success) {
      return res.data.data.translateText;
    }
    return '';
  } catch (error) {
    console.error('Alibaba爬虫失败：', error);
    return '';
  }
}
